### ISSUE TITLE
`_Static_assert` after case label is not recognized

### ISSUE THREAD

**Issue by zufuliu** [llvm/llvm-project#55821]
Copied from https://github.com/mstorsjo/llvm-mingw/issues/285, online https://godbolt.org/z/nsPEh54h6

See build errors at https://github.com/zufuliu/notepad2/runs/6699313318?check_suite_focus=true
reproduce:
```c
#if !defined(__cplusplus)
#define static_assert(expr)		_Static_assert(expr, #expr)
#endif

int test(int id) {
	static_assert(1 == 1);
	switch (id) {
	case 1:
		static_assert(1 == 1);
		return 1;
	default:
		return 0;
	}
}
```

```
D:\notepad2\notepad2\src>x86_64-w64-mingw32-clang++ -c -xc++ -std=c++17 test.c

D:\notepad2\notepad2\src>x86_64-w64-mingw32-clang -c -std=c17 test.c
test.c:9:3: error: expected expression
                static_assert(1 == 1);
                ^
test.c:2:30: note: expanded from macro 'static_assert'
#define static_assert(expr)             _Static_assert(expr, #expr)
                                        ^
1 error generated.
```

This may be an upsteam llvm issue for mingw target, with official 14.0.4 binary:
```
D:\notepad2\notepad2\src>clang++ --target=x86_64-pc-windows-msvc -c -xc++ -std=c++17 test.c

D:\notepad2\notepad2\src>clang++ --target=x86_64-pc-windows-gnu -c -xc++ -std=c++17 test.c

D:\notepad2\notepad2\src>clang --target=x86_64-pc-windows-msvc -c -std=c17 test.c

D:\notepad2\notepad2\src>clang --target=x86_64-pc-windows-gnu -c -std=c17 test.c
test.c:9:3: error: expected expression
                static_assert(1 == 1);
                ^
test.c:2:30: note: expanded from macro 'static_assert'
#define static_assert(expr)             _Static_assert(expr, #expr)
                                        ^
1 error generated.
```

**Comment by mstorsjo** [comment]
> This may be an upsteam llvm issue for mingw target

The issue isn't specific to the mingw target, it works the same with e.g. `--target=x86_64-linux-gnu` too (https://godbolt.org/z/nsPEh54h6 runs without any specific `--target` set). So apparently this is just handled differently in MSVC mode compared to the rest.

GCC also seems to have errored out on this before; with GCC 9 and 10, I get a similar error:
```
$ gcc -std=c17 -c test.c
test.c: In function ‘test’:
test.c:2:30: error: a label can only be part of a statement and a declaration is not a statement
    2 | #define static_assert(expr)  _Static_assert(expr, #expr)
      |                              ^~~~~~~~~~~~~~
test.c:9:3: note: in expansion of macro ‘static_assert’
    9 |   static_assert(1 == 1);
      |   ^~~~~~~~~~~~~
```
However, starting with GCC 11, the construct seems to be accepted.

**Comment by zufuliu** [comment]
Oh, add brace after colon indeed works.
```cpp
#if !defined(__cplusplus)
#define static_assert(expr)		_Static_assert(expr, #expr)
#endif

int test(int id) {
	static_assert(1 == 1);
	switch (id) {
	case 1: {
		static_assert(1 == 1);
		return 1;
	}
	default:
		return 0;
	}
}
```

**Comment by DimitryAndric** [comment]
Yeah this seems to fail to build mostly everywhere, for example on FreeBSD with clang 14.0.3:

```c
int test(int id) {
        _Static_assert(1 == 1, "foo");
        switch (id) {
        case 1:
                _Static_assert(1 == 1, "bar");
                return 1;
        default:
                _Static_assert(1 == 1, "baz");
                return 0;
        }
}```

gives:

```
% clang -std=c17 -c pr55821.c
pr55821.c:5:3: error: expected expression
                _Static_assert(1 == 1, "bar");
                ^
pr55821.c:8:3: error: expected expression
                _Static_assert(1 == 1, "baz");
                ^
2 errors generated.
```

For many older compilers, _Static_assert was defined with something like:

```
#define _Static_assert(x, y)  typedef char __assert_ ## y[(x) ? 1 : -1] __unused
```

so maybe clang is doing something similar internally, and expects to be able to insert a declaration at that point?

**Comment by AaronBallman** [comment]
The standards references and citations are a bit strange because the grammar productions changed between C17 and C2x in this area, but the results are the same.

C17 6.7p1 says that `_Static_assert` is a kind of `declaration`. C17 6.8.1p1 says that each kind of `labeled-statement` must be followed by a `statement`. C17 6.8p1 has no production which allows for a `declaration` to appear as a statement, but C17 6.8.2p1 for `compound-statement` has a `block-item-list` production which does include `declaration`. So Clang is correct to diagnose this as an error -- the syntax of the language doesn't allow for it.

C2x has the same behavior, but it split up the productions into `label` and `labeled-statement`, and added the notion of an `unlabeled-statement` as well. But what follows the production is still a `statement` and there's still no ability to use an arbitrary `declaration` from there, so Clang should produce the same results in C2x.

This is different than the grammar in C++, which allows a `declaration-statement` within the `statement` production (http://eel.is/c++draft/stmt.pre#nt:statement), so Clang is correct to accept that code in C++.

So unfortunately, I think this is just one of those weird edge case differences between the languages but isn't actually a compiler bug. It could be argued that this might be a place where Clang might want to issue a diagnostic in pedantic mode but otherwise accept the code. This is the route that GCC went: https://godbolt.org/z/1MGxT1o7G

**Comment by zufuliu** [comment]
Then this is a c2x feature as per GCC 11 release note from https://gcc.gnu.org/gcc-11/changes.html
> * Several new features from the upcoming C2X revision of the ISO C standard are supported with `-std=c2x` and `-std=gnu2x`. Some of these features are also supported as extensions when compiling for older language versions. In addition to the features listed, some features previously supported as extensions and now added to the C standard are enabled by default in C2X mode and not diagnosed with `-std=c2x -Wpedantic`. 
>   * Labels may appear before declarations and at the end of a compound statement.

No warnings with `gcc -std=c2x -pedantic`, https://godbolt.org/z/qnrY77ae4

**Comment by zufuliu** [comment]
C23 latest working draft http://open-std.org/JTC1/SC22/WG14/www/docs/n2731.pdf

> ### 6.8.2 Compound statement
> Syntax
> ```
> 1 compound-statement:
> 	{ block-item-listopt }
> block-item-list:
> 	block-item
> 	block-item-list block-item
> block-item:
> 	declaration
> 	unlabeled-statement
> 	label
>  ```
> Semantics
> 
> 2 A compound statement is a block. A label shall be translated as if it were followed by a null statement.

**Comment by AaronBallman** [comment]
@zufuliu -- you're correct! This was [WG14 N2508](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n2508.pdf) which I had found earlier but misread the grammar change to block-list. We don't currently claim to [implement](https://clang.llvm.org/c_status.html#c2x) that one yet, but I'm sure we'll get to it!

**Comment by zufuliu** [comment]
A related C++23 feature regrading label and compound statement, https://gcc.gnu.org/gcc-13/changes.html

> [P2324R1](https://wg21.link/p2324), Labels at the end of compound statements ([PR103539](https://gcc.gnu.org/PR103539))

**Comment by zufuliu** [comment]
Maybe this can be closed as duplicate of issue #56040.

**Comment by AaronBallman** [comment]
Closing as a duplicate of https://github.com/llvm/llvm-project/issues/56040
